### ISSUE TITLE
High CPU usage when there are at least two failed tasks on the queue 

### ISSUE THREAD

**Issue by jifang** [yigit/android-priority-jobqueue#262]
Hi, we observed a very high CPU usage, similar to issue #80 , if we added at least two network tasks to the queue which can't be completed, for example, request timeout. 

To reproduce, simply schedule two tasks with invalid URL. For example, `http://www.example.com:8080`

The sample project can be found [here](https://github.com/jifang/jobqueuetest)

We are using the V2

```
10-07 14:03:57.478 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237478166641
10-07 14:03:57.478 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237478166641
10-07 14:03:57.478 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.478 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.478 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message null
10-07 14:03:57.478 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_ADD_JOB_2] remove message null
10-07 14:03:57.478 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_JOB_CONSUMER_IDLE_5] remove message null
10-07 14:03:57.478 30114-30140/co.visbit.testjobqueue D/JOBS: joq idle. running:? true
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: Job queue idle. next job at: 1475874233744962894
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: add delayed message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866 at time 1475874233744962894
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] did on idle post a message? true
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237479977891
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237479977891
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] post message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] consuming message of type CONSTRAINT_CHANGE
10-07 14:03:57.480 30114-30140/co.visbit.testjobqueue D/JOBS: considering adding a new consumer. Should poke all waiting? true isRunning? true waiting workers? 0
10-07 14:03:57.481 30114-30140/co.visbit.testjobqueue D/JOBS: check above load factor: totalCons:1 minCons:1 maxConsCount: 3, loadFactor 3 remainingJobs: 1 running holders: 1. isAbove:false
10-07 14:03:57.481 30114-30140/co.visbit.testjobqueue D/JOBS: nothing has been poked. are we above load factor? false
10-07 14:03:57.481 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237481594505
10-07 14:03:57.482 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237481594505
10-07 14:03:57.482 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.482 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.482 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message null
10-07 14:03:57.482 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_ADD_JOB_2] remove message null
10-07 14:03:57.482 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_JOB_CONSUMER_IDLE_5] remove message null
10-07 14:03:57.482 30114-30140/co.visbit.testjobqueue D/JOBS: joq idle. running:? true
10-07 14:03:57.483 30114-30140/co.visbit.testjobqueue D/JOBS: Job queue idle. next job at: 1475874233744962894
10-07 14:03:57.483 30114-30140/co.visbit.testjobqueue D/JOBS: add delayed message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866 at time 1475874233744962894
10-07 14:03:57.483 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] did on idle post a message? true
10-07 14:03:57.484 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237483677422
10-07 14:03:57.484 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237483677422
10-07 14:03:57.484 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] post message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.484 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.484 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.484 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.484 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] consuming message of type CONSTRAINT_CHANGE
10-07 14:03:57.484 30114-30140/co.visbit.testjobqueue D/JOBS: considering adding a new consumer. Should poke all waiting? true isRunning? true waiting workers? 0
10-07 14:03:57.485 30114-30140/co.visbit.testjobqueue D/JOBS: check above load factor: totalCons:1 minCons:1 maxConsCount: 3, loadFactor 3 remainingJobs: 1 running holders: 1. isAbove:false
10-07 14:03:57.485 30114-30140/co.visbit.testjobqueue D/JOBS: nothing has been poked. are we above load factor? false
10-07 14:03:57.485 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237485412057
10-07 14:03:57.485 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237485412057
10-07 14:03:57.485 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.485 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.485 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message null
10-07 14:03:57.486 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_ADD_JOB_2] remove message null
10-07 14:03:57.486 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_JOB_CONSUMER_IDLE_5] remove message null
10-07 14:03:57.486 30114-30140/co.visbit.testjobqueue D/JOBS: joq idle. running:? true
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: Job queue idle. next job at: 1475874233744962894
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: add delayed message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866 at time 1475874233744962894
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] did on idle post a message? true
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237487221797
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237487221797
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] post message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] consuming message of type CONSTRAINT_CHANGE
10-07 14:03:57.487 30114-30140/co.visbit.testjobqueue D/JOBS: considering adding a new consumer. Should poke all waiting? true isRunning? true waiting workers? 0
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: check above load factor: totalCons:1 minCons:1 maxConsCount: 3, loadFactor 3 remainingJobs: 1 running holders: 1. isAbove:false
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: nothing has been poked. are we above load factor? false
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237488842995
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237488842995
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message null
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_ADD_JOB_2] remove message null
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_JOB_CONSUMER_IDLE_5] remove message null
10-07 14:03:57.489 30114-30140/co.visbit.testjobqueue D/JOBS: joq idle. running:? true
10-07 14:03:57.490 30114-30140/co.visbit.testjobqueue D/JOBS: Job queue idle. next job at: 1475874233744962894
10-07 14:03:57.490 30114-30140/co.visbit.testjobqueue D/JOBS: add delayed message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866 at time 1475874233744962894
10-07 14:03:57.490 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] did on idle post a message? true
10-07 14:03:57.491 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237490645860
10-07 14:03:57.491 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237490645860
10-07 14:03:57.491 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] post message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.491 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.491 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.491 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.491 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] consuming message of type CONSTRAINT_CHANGE
10-07 14:03:57.491 30114-30140/co.visbit.testjobqueue D/JOBS: considering adding a new consumer. Should poke all waiting? true isRunning? true waiting workers? 0
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: check above load factor: totalCons:1 minCons:1 maxConsCount: 3, loadFactor 3 remainingJobs: 1 running holders: 1. isAbove:false
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: nothing has been poked. are we above load factor? false
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237492243620
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237492243620
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message null
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_ADD_JOB_2] remove message null
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_JOB_CONSUMER_IDLE_5] remove message null
10-07 14:03:57.492 30114-30140/co.visbit.testjobqueue D/JOBS: joq idle. running:? true
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: Job queue idle. next job at: 1475874233744962894
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: add delayed message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866 at time 1475874233744962894
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] did on idle post a message? true
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237494328412
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237494328412
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] post message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.494 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.495 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] consuming message of type CONSTRAINT_CHANGE
10-07 14:03:57.495 30114-30140/co.visbit.testjobqueue D/JOBS: considering adding a new consumer. Should poke all waiting? true isRunning? true waiting workers? 0
10-07 14:03:57.496 30114-30140/co.visbit.testjobqueue D/JOBS: check above load factor: totalCons:1 minCons:1 maxConsCount: 3, loadFactor 3 remainingJobs: 1 running holders: 1. isAbove:false
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: nothing has been poked. are we above load factor? false
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237496756485
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237496756485
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message null
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_ADD_JOB_2] remove message null
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_JOB_CONSUMER_IDLE_5] remove message null
10-07 14:03:57.497 30114-30140/co.visbit.testjobqueue D/JOBS: joq idle. running:? true
10-07 14:03:57.498 30114-30140/co.visbit.testjobqueue D/JOBS: Job queue idle. next job at: 1475874233744962894
10-07 14:03:57.498 30114-30140/co.visbit.testjobqueue D/JOBS: add delayed message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866 at time 1475874233744962894
10-07 14:03:57.498 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] did on idle post a message? true
10-07 14:03:57.498 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] looking for next message at time 1475874237498610912
10-07 14:03:57.499 30114-30140/co.visbit.testjobqueue D/JOBS: flushing messages at time 1475874237498610912
10-07 14:03:57.499 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] post message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.499 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] next delayed job null
10-07 14:03:57.499 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_RUN_JOB_RESULT_6] remove message null
10-07 14:03:57.499 30114-30140/co.visbit.testjobqueue D/JOBS: [queue_CONSTRAINT_CHANGE_4] remove message com.birbit.android.jobqueue.messaging.message.ConstraintChangeMessage@8102866
10-07 14:03:57.499 30114-30140/co.visbit.testjobqueue D/JOBS: [priority_mq] consuming message of type CONSTRAINT_CHANGE
10-07 14:03:57.499 30114-30140/co.visbit.testjobqueue D/JOBS: considering adding a new consumer. Should poke all waiting? true isRunning? true waiting workers? 0
```

**Comment by yigit** [comment]
thanks for the sample. I'll check it!

**Comment by yigit** [comment]
hmm interesting, next job to be run timeout returns a negative number :/. So job manager tries to re-runs instantly but it does not find any job to be run :/. Need to disconnect now but will try to check this tonight. Thanks a lot for the report, I'll try to fix and release asap.

**Comment by yigit** [comment]
another interesting thing, if I replace the job contents w/ a simple `throw RuntimeException`, it does not reproduce.

**Comment by yigit** [comment]
Ok i think I found the problem and why 2 jobs trigger it.

When JobManager becomes idle, it asks job queues to get the timestamp of the next available job so that it can wake up at that moment.
In your case, the queue returns `now` because it has 2 jobs and only 1 of them runs due to the configuration (the load factor is 3, meaning that as long as waiting job count is less than 3, we won't create a new thread).
So when JobManager tries to decide when the next job is available, the answer is `now` but when it tries to consume it, JobManager cannot run it because the consumers are full and then this creates an infinite loop until after the first 10 seconds (the network failure in my case takes 10 seconds). Eventually as jobs back off, this will settle but that will take a long time.
It is not a trivial problem so there is no instant fix but i'll work on it.

**Comment by yigit** [comment]
fix released in 2.0.1. Thanks.
